# Creating tables of **letters pairs**

##About
There is a task of decoding "crypted" text, which is based on the number of occurrences of pairs of letters in "natural language text".

---
Є задача декодування "зашифрованого" тексту, яка базується на кількості частоти входжень пар літер в "natural language text".

Цей ноутбук призначений беспосередньо для рішення поставленої задачі: виявити частоту зучтрічей пар літер в тексті.

Порядок дій наступний:
- Читаємо вміст оброблених текстових файлів у словник.
- Створюємо кількісні таблиці по кожному файлу і для всього тексту в цілому.
- Створюємо кількісні та частотні таблиці по всіх текстів враховуючи пробіл та без нього.
- Створюємо `зміщені` кількісні та частотні таблиці по всіх текстів враховуючи пробіл та без нього.

Статистика по ціх табліцях буде в іншому ноутбуці. Це пов'язано з тим, що створення кількісних таблиць по текстових файлах займає багато часу. Тому цей ноутбук використовується "одноразово", тільки для створення таблиць. У подальшому процесі всі створені таблиці читаються з диска (google drive).

The following texts were used:
- (1 HS) "Hamlet" by Shakespeare ([here](https://shakespearestudyguide.com/Hamlet%20Text.html)) [Without expressions like "BERNARDO:"]
- (2 TB) Torah, Sefer Bereshit in English ([here](https://www.tanach.us/Pages/About.html))
- (3 AR) Asimov "I, Robot"  ([here](https://royallib.com/book/Asimov_Isaac/I_Robot.html))
- (4 HP) "Harry Potter and the Philosopher's Stone" ([here](https://github.com/amephraim/nlp/blob/master/texts/J.%20K.%20Rowling%20-%20Harry%20Potter%201%20-%20Sorcerer's%20Stone.txt))
- (5 GG) Genetics : From Genes to Genomes. 6th edition ([here](http://skgjx.whu.edu.cn/Public/upfile/article/202103031656469260.pdf)) [Intro & parts 1,2,4. My edit from PDF]
- (6 SA) Scientific articles: from IEEE, [medium.com](https://medium.com/) etc.
- (7 NA) News articles: [New York Times](https://www.nytimes.com/international/), [Fox News](https://www.foxnews.com/), [BBC](https://www.bbc.com/news) etc.
- (8 CC) Coments in chats








---
---
## Import & mount

In [1]:
import os
import re
import copy
from time import time
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
from matplotlib import cm
from mpl_toolkits.mplot3d import Axes3D
import seaborn as sns

# from matplotlib import pyplot as plt
# plt.rcParams['figure.figsize'] = [15, 6]

In [2]:
# Mount GitHub
!git clone https://github.com/EdwardGerman/Columnar-Transposition-Cipher.git  # clone repository
%ls  # checking whether all files are present
drch = '/content/Columnar-Transposition-Cipher' # Path to data

folder_pp = 'Data_pp'
folder_pt = 'Parity_tables'

Mounted at /content/drive


In [3]:
title_font = {'family': 'serif',
              'color':  'darkred',
              'weight': 'bold',
              'size': 14,
              }

---
---
## Funcfions

#### Space_n_letters()
Function for count of all letters and spaces in preproc text

In [4]:
def Space_n_Letters(pp_text, p = False):
    chars_after_PP = len(pp_text)
    spaces = pp_text.count(' ')
    letters = chars_after_PP - spaces

    if p == False:
        return [chars_after_PP, letters, spaces]
    else:
        print('Symbols number:', chars_after_PP)
        print('Spaces number:', spaces)
        print('Only letters: ', letters,'\n')
    # print('Only letters (2): ', len(list(filter(str.isalpha, pp_text))),'\n')

#### CharacterCounts()
Counts the number of each character in the text and print sorted [by decreasing] list (dictionary).

!Noties! optiotal:
* Upper letter = lower letter
* Multyspace -> space

In [5]:
# Returns sorted dictionary {letter : count} from largest to smallest number
def CharactersCount(orig_text, multispace = False, uppercase = False):
    text = copy.copy(orig_text)
    if multispace == False: text = re.sub(r' +', ' ', text)
    if uppercase  == False: text = text.lower()

    counts = {}   # Creating empty dictionary
    for char in text:    # Loop for count characters
        if char in counts: counts[char] += 1
        else:              counts[char]  = 1
    return dict(sorted(counts.items(), key=lambda item: item[1], reverse=True))

#### Display of the DataFrames
Function for better show of the DataFrames

In [6]:
def display_df(df, name = True):
    pd.set_option('display.max_columns', None) # Print all rows (features) in DF
    if name == True and hasattr(df, 'name'): print(df.name + ':')
    display(df)
    print('\n')
    pd.reset_option('display.max_columns')     # Default setting: print print only first & final 5 rows


---
---
---
## Creating ***quantitativity tables***
Order is following:
1. Make a `dictionary` with texts and adding a *united text*
    * Make files list and checking files
    * Read texts from dict
    * Add to dict united text
2. Compiling a `quantitativity tables` for pairs
    * Creating function, what compiled *quantitativity tables*
    * Compiling quantitativity tables by each text including united text


---
### Make a ***dictionary*** with texts and adding a *united text*

Make files list and checking files

In [7]:
# Get list of all files in directory
file_list = os.listdir(os.path.join(drch, folder_pp))

# Checking:
for name in file_list:
    if not name.endswith('_pp.txt'):
        print(name, ' - error pattern: not ended on "_pp.txt"')

# Get list of key names for dictionary
file_names = [name.replace('_pp.txt', '') for name in file_list if name.endswith('_pp.txt')]
display(file_names)

['Torah_Bereshit',
 'Genetics_124',
 'Chat_Comments',
 'Asimov_I_Robot',
 'Harry_Potter_I',
 'News_articles',
 'Hamlet_BB',
 'Sci_articles',
 'All texts']

***Read*** text to dict and ***adding*** united text:

In [8]:
# Read text to dict:
texts = {}  # Create empty dict

# Loop for each file:
for text_file in file_list:

    # Read:
    path_r = os.path.join(drch, folder_pp, text_file)
    with open(path_r, 'r', errors='ignore') as file:
        text = file.read()

    texts[text_file.replace('_pp.txt', '')] = text


### Creating ***quantitative*** tables

Function for compiled `quantitative` tables for each text files

In [9]:
def quantitative_table(name, text):
    # Creating names list for index & columns of DF: 'space' + all letters from 'a' to 'z'
    all_letters = [' '] + [chr(i) for i in range(ord('a'), ord('z') + 1)]

    df = pd.DataFrame(index=all_letters, columns=all_letters)   # call the indexes and columns
    df = df.fillna(0)                                           # reset to zero all values in DF


    # Iterate by text
    for i in range(len(text) - 1):
        first = text[i]
        second = text[i + 1]
        # Increase by 1 value of cell with corresponding row (index) and column
        df.at[first, second] += 1

    df.index = [s + '-' for s in all_letters]
    df.columns = ['-' + s for s in all_letters]
    df.name = name
    return df

Compiling `quantitative` tables for each text

In [10]:
quantit_tables = {}

for name in texts:
    quantit_tables[name] = quantitative_table(name, texts[name])

    # Save DataFrame with quantitative Parity Table to disk (Goodle drive)
    path_pt = os.path.join(drch, folder_pt, name + '.csv')
    quantit_tables[name].to_csv(path_pt, index_label=False)


Save DataFrames with *`quantitative`* `Parity Tables` to disk (Goodle drive)

It could be done in the previous for-loop, but for better understanding, I decided to do it separately.

In [11]:
# for name in quantit_tables:
#     path_pt = os.path.join(drch, folder_pt, name + '.csv')
#     quantit_tables[name].to_csv(path_pt, index_label=False)

## Statistics

Quantitative parity table for `All texts`:

In [12]:
print('All chars:' ,quantit_tables['All texts'].values.sum())
display_df(quantit_tables['All texts'])

All chars: 2263891
All texts:


-      -a     -b     -c     -d     -e     -f     -g     -h     -i  \
 -      0  43411  17690  19257  14681   9638  16856  10019  24691  28471   
a-  11562    276   3297   6124   4707    280    952   2996    718   6242   
b-   1036   2469    358     40     13   7706      3     19      9   1760   
c-   2372   7287     21    715     41   7201     53    183   8266   2405   
d-  42369   2368     98    120    759   9265    129    295     23   6460   
e-  79473  11005    366   5636  17124   6191   1867   1576    602   2720   
f-  14427   2479      2     21      9   3658   2054     24      4   4152   
g-  13712   2424     34     64     36   6488     12    505   3845   1958   
h-  10502  18077     28     20     94  45248     26      4     24  12640   
i-   4618   2483   1038   7357   6508   4401   3101   4233     63    252   
j-    170    410      1      3      0    384      0      0      0     46   
k-   4642    302     19     31    113   4470     25     66     59   1855   
l-  13499   6647     72    257   4138  14720    960    126     39   7701   
m-   8423   8649   1345    198     22  11791     73     23      3   4390   
n-  32010   4257     81   4293  18986  11040    658  15023    284   4093   
o-  17262    944   2298   1732   3133    576  12614   1393    425   1581   
p-   3343   3588     42    106     52   6856     16     14   1478   1652   
q-    229      6      0      0     17     12      7      2      1      8   
r-  23823   8216    383    978   2993  25057    480   1151    156   8521   
s-  51841   4668     91   1577     88  13383    195    113   4787   7469   
t-  41282   6563     94   1037     30  15714    157     18  50334  14066   
u-   4601   1284    888   1987   1397   1746    332   2158     29   1409   
v-    530   1820      4     15     61  12326      2      3      6   2810   
w-   4673   6504     15     45     98   6493     22      1   6203   6380   
x-   1187    428     32    287     10    241      5     14     24    436   
y-  23506    148    236    139     80   2470    108    186      7    731   
z-    219    258      1      5      6    860      1      0     22    208   

      -j    -k     -l     -m     -n     -o     -p    -q     -r     -s     -t  \
 -  2121  3059  10513  16610   9263  24615  15370  1162  10588  30310  61470   
a-    98  2080  13634   5119  27631    135   2735    23  15403  12045  20611   
b-   139     1   3439     14     47   3821     30     0   2280    442    170   
c-     0  2674   1932     28     21   9134     14    20   2215    483   4608   
d-    23    11    667    151   1009   4604     23    85   1668   1848     61   
e-    47   310   8886   4734  18902   1046   2516   635  28192  17631   7256   
f-     0     1   1234      1      1   7167     24     0   2661    111   1078   
g-     1     4    924    222    803   3591     97     5   2991    789    153   
h-     0     5    163    208    238   7340     16     3   2102    244   2589   
i-    37  1085   6260   5267  34002   8238   1019   145   4347  15540  15180   
j-     1     1      0      0      1    626     13     1     12      2      0   
k-     8    10    318     14   1288     77     95     0     53    689     14   
l-     0   400  11627    296     61   6975    254     1    150   2020   1399   
m-     1     4    265    698    125   5360   2742     1    245   1296     27   
n-    96  1275   1064    257   1116   8393    127    26     63   6001  14299   
o-    69  1716   5278   8905  22284   4531   3153     5  16974   5155   7866   
p-     1    10   4461    135     10   4555   2109     1   5136    731   1122   
q-     0     0      0      0      1      1      7     0      0      6     10   
r-     5   978    944   2663   2115  13075    514    98   3330   5633   4143   
s-    39   642   1021    885    794   6107   2829   121    196   4599  13966   
t-     2     3   1514    183    116  14715    131     1   5702   5474   2500   
u-     5    88   4968   2072   5276     96   2108     6   7331   6748   6712   
v-     0     0     10     25     23   

Quantitative parity table for each text:

In [13]:
for table in quantit_tables:
    print('-'*60, '\nQuantitative table for "', table, '"',sep = '')
    print('All chars:' ,quantit_tables[table].values.sum())
    display_df(quantit_tables[table], name = False)
    print('\n\n')

------------------------------------------------------------
Quantitative table for "Torah_Bereshit"
All chars: 181244


-     -a    -b    -c    -d    -e    -f   -g    -h    -i   -j   -k    -l  \
 -     0  3859  1784  1049   974   830  1488  900  3129  2055  453  219  1206   
a-   419   148   368   582   399    72   141  122   511   791    0  196   716   
b-   195   171     2     0     2   906     0    0     0    99    0    0   141   
c-    78   560     0    40     0   333     0    0   523    96    0  172    36   
d-  5412   357     3     0    28   391     2   22     0   267    0    0    21   
e-  7376  1001    96   266  1390   425   209  153   154   287    0   49   696   
f-  1547   386     0     0     0   240    56    0     0   191    0    0   117   
g-   581   210     0     0     3   200     0    3   384   128    0    0    19   
h-  1534  1701     5     2     1  5802     0    0     6  1739    0    1     9   
i-   493    49    23   214   680   305   267  198     1     5    0   29   763   
j-     0   230     0     0     0    14     0    0     0     2    0    0     0   
k-   293    37     0     1     0   366     0    0     2   135    0    1    12   
l-  1248   589     0     8   441   849    61    2    10   412    0   22  1085   
m-  1316   634    46    18     1  1023     7    0     0   158    0    0     3   
n-  2515   334     3   105  2526   620     3  711    12   197   18   57    32   
o-  1674   111   210   109   441    11  1409   22   116    60    0  182   294   
p-   245   128     8     0     0   262     0    0   351    84    0    0   221   
q-     0     0     0     0     0     0     0    0     0     0    0    0     0   
r-  2365   757     9    34   455  1786    18   25     5   361    0   14    22   
s-  3703   964     9    27     0  1361     0    0   640   205    0   49    22   
t-  2652   236    20    20     0   817    16    0  5632   310    0    0    70   
u-   714    24    42    40    58    32     2  246     2    62    2    2   170   
v-     2   133     0     0     0  1235     0    0     0   116    0    0     0   
w-   338   560     0     0     1   591     0    0   600   946    0    0     3   
x-     8     2     0    16     0     0     0    0     0     0    0    0     0   
y-  1621     2     1     0     0   215     0    0     0   101    0    0     0   
z-    22    11     1     0     0    30     0    0     0    24    0    0     2   

      -m    -n    -o   -p  -q    -r    -s    -t    -u   -v    -w  -x    -y  -z  
 -  1498   614  2158  665   8   417  2745  6038   305   92  2554   0  1263  48  
a-   854  3120    99  152   0  1092   837  1315   295  376   136   0   436  17  
b-     1     0   266    0   0   535    33     5   187    0     0   0    86   1  
c-     0     0   530    0   0    64     5    16    77    0     0   0     1   0  
d-    32    72   312    0   0   244   159     3    34    1    18   0    21   0  
e-   376  1260    88  324   8  2357   974   408    29  297    70  33   375  15  
f-     0     0   639    0   0   253    18   179    52    0     0   0     1   0  
g-     0    39   497    0   0   166    45     7    26    0     0   0    97   0  
h-    23     0   577    7   0    94    67   344   132    2     2   0    29   1  
i-   603  1790   132   69   3   360  1431   952     1  426     0  18     0  19  
j-     0     0   181    0   0     0     0     0    46    0     0   0     0   0  
k-     0    67     6    0   0     0    49     3     0    0     3   0    18   0  
l-    15     1   466   13   0     0   149    71    19   51     0   0   148   0  
m-    37     9   251   22   0    16    83     0    40    0     0   0   368   0  
n-     1    12   424    0   1     0   379   773    13    3     0   0    44   0  
o-   497  1183   294   76   0  1213   290   489  1837  163   358   0    38   1  
p-     1     0   125  103   0   153    15   146    55    0     2   0     2   0  
q-     0     0     0    0   0     0     0     0    20    0     0   0     0   0  
r-    36   163   650   29   0    54   540   251    71  129     6   0   184   0  
s-    19    17   561  154   0    47   276   568    89    0    51   0     8   0  
t-     0    12  1473    0   0   160 






------------------------------------------------------------
Quantitative table for "Genetics_124"
All chars: 379157


-     -a    -b    -c    -d    -e    -f    -g    -h    -i  -j   -k  \
 -      0  6998  2294  4728  2586  1988  2940  2093  2081  4352  76  262   
a-   2201    31   478   971   341    10   171   271     2   906   7  139   
b-    232   323    43     0     4   975     0     0     1   439  29    0   
c-    688  1398    12   167     1  1878    49    10  2332   586   0  162   
d-   5016   197    11     4    69  1681     5    31     1  1803   2    0   
e-  12977  1557    35  1438  2678   787   182   415    42   722   5   21   
f-   3184   248     0     1     0   841   451     0     0   719   0    0   
g-   1889   522     1     6     0  2011     8   120   352   309   0    0   
h-   1603  2366     9     0    39  6467     0     1     1  1142   0    1   
i-    386   499   286  1803   987   777   602   677    41   193   0  117   
j-      7    15     0     1     0    57     0     0     0     2   0    0   
k-    273    78     9     0     4   380     0     2     0   183   0    0   
l-   2113  1123    18    11   445  3269    90     3     1  1277   0   14   
m-    870  2133   300    11     5  2708     6     0     0  1019   0    0   
n-   5666  1192     9   932  2864  2413   210  1977   168   759   7  186   
o-   2045    62   323   422   553    64  2882   485    89   167  15   95   
p-    266   856     1     8     3  1503     0     1   525   285   0    3   
q-      3     0     0     0     0     0     0     0     1     0   0    0   
r-   3512  1538    12   124   193  4297    52   303    15  1686   0  106   
s-  10457   218    21   359     2  2825    17    71   439  1943  26   67   
t-   4438  1177     7   103     2  3433     4     8  7734  3636   0    0   
u-    212   401   142   656   168   307    25   250     0   165   0   21   
v-     30   307     0     0     0  1700     2     0     0   681   0    0   
w-    694   345     2     1    18   559     0     0   756   831   0    0   
x-    562   128     0    60     0    32     0    10    13    61   0    0   
y-   3159     7   102    27    11   389     2   176     0    62   0    0   
z-      8    73     0     0     0   197     0     0     1    41   0    0   

      -l    -m    -n    -o    -p   -q    -r    -s    -t    -u   -v    -w   -x  \
 -   977  2630  1200  4875  3261   64  1845  3649  9104   559  332  2287  452   
a-  3288   689  4277     2   607    2  2484  1761  4115   356  348   111   13   
b-   469     0    29   413     2    0   447    95    29   254    3     0    2   
c-   407     1     3  1690     0    3   567   160  1135   389    0     0    0   
d-    80     6   197   426     8    0   228   384     2   682   12    12    0   
e-  1978   758  4823   176   428  208  4338  4587  1511    83  566   131  793   
f-   169     0     0  1188    16    0   472    61    99   209    0     1    0   
g-   128    37   122   380     0    0   532   148    68   217    0     1    0   
h-    39    13    56  1000     2    0  1046    26   189   351    0    21    0   
i-   972   471  6329  2396   196   26   612  3135  2173    11  965     1   35   
j-     0     0     0    28     1    0     0     0     0    56    0     0    0   
k-    64     0   115     1    14    0     3    47     3     5    0     3    0   
l-  2094    18     3  1289    30    0     7   623   314   292   79    29    0   
m-     6    85    32  1627   479    0     8   282     0   272    1     0    3   
n-   146    40   143  1327     4    2    17  1296  2530   313   93     3    1   
o-  1063  3362  3841   240   624    0  2744  1924  1370  1528  311   988   58   
p-   991    45     1   594   296    0  1492    77   244   225    0     1    0   
q-     0     0     0     0     0    0     0     0     1   338    0     0    0   
r-   117   863   354  3375   119    0   400   832   595   213  110    42    3   
s-    38   292    10  1434   573   37    52   963  1782   586    1    61    0   
t-   124    35     4  1895     4    0  1015   963   294   437    0   470    0   
u-   938   432   793    24   191    1   959  1073   659     0    6     1    1   
v-     0     0 






------------------------------------------------------------
Quantitative table for "Chat_Comments"
All chars: 277917


-     -a    -b    -c    -d    -e    -f    -g    -h    -i   -j   -k  \
 -     0  5632  2349  2332  1628  1226  1906  1192  2030  4118  397  316   
a-  1582     7   539   627   544    81   114   456    57   472   29  285   
b-   119   348    20     7     3  1045     2     1     6   281    7    1   
c-   274  1090     4   108     5   781     1     3   735   371    0  355   
d-  3972   251     4    14    98  1170     8    28     3   812    0    0   
e-  9738  1321    83   721  1478   692   189   168    29   463    3   34   
f-  1415   315     0     6     3   361   248    14     0   335    0    1   
g-  1965   322     7     4     5   807     0    46   386   231    0    0   
h-  1033  2137     4     7     3  4569     6     0     6  1436    0    3   
i-   769   435   138  1046   612   560   386   442     4     7    1  246   
j-    13    29     1     0     0    56     0     0     0     6    1    0   
k-   618    43     3     0     7   675     7     4     4   245    0    4   
l-  1617   731     5    15   625  1660    66    17    12  1289    0   74   
m-   881   950   120    16     1  1595     7     4     3   557    1    2   
n-  3638   503    13   409  2061  1427    68  2170    23   523   13  203   
o-  2509   111   201   246   303   134  1179   131    48   208   13  224   
p-   446   479     9    19     5   862     4     2    91   223    0    0   
q-    13     6     0     0     0     3     0     0     0     2    0    0   
r-  2836  1003    38   106   324  3353    26   147    22  1128    1  155   
s-  6858   509     8   169    11  1402    30    11   576   779    2   56   
t-  5584   804    31    98     5  1815    30     6  6137  1883    1    0   
u-   496   146   183   298   202   265    45   252     5   140    1   40   
v-    35   163     0     0     0  1682     0     0     0   374    0    0   
w-   589   819     3    33     6   756    12     1   819   682    0    2   
x-   109    45    29    41     3    16     0     2     0    37    1    0   
y-  3208    38    40    17     8   247     0     4     2    97    1    0   
z-    36    20     0     0     1    85     1     0     1    54    0    1   

      -l    -m    -n    -o    -p  -q    -r    -s    -t    -u   -v    -w   -x  \
 -  1443  1953  1265  2872  2101  95  1312  3243  7723   617  398  3180   72   
a-  1889   641  3514     9   269  12  1858  1355  2345   311  530   128   18   
b-   486     5     5   529     0   0   149   106    23   513   13     0    0   
c-   235     0     0  1259     2   4   303    63   507   182    2     0    0   
d-    62    17    78   634     6   0   173   227     5   231   26    16    0   
e-   925   656  2087   351   298  42  3264  2376   791    58  487   210  279   
f-    66     0     0   930     4   0   310     3    88   220    0     4    0   
g-    73    17   104   492     1   5   325   123     9   137    1     1    1   
h-    32    10    48  1050     3   0    92    33   261   110    0     2    0   
i-   803   651  4134   819   115  19   618  2266  2146    14  412     4   16   
j-     0     0     0   138     2   1     2     1     0   220    2     0    0   
k-    29     1   139    17     2   0    31   132     3    10    0    10    1   
l-  1383    59    11   765    39   0    30   261   178   223   54    39    0   
m-    18   104    14   635   368   0     4   128     4   188    0     8    1   
n-   159    72   172  1045    13   4    11   821  1814    90   84    25    2   
o-   686   951  2874   528   588   1  2267   494  1073  2210  360   783   48   
p-   649    14     0   700   275   0   671   142   102   212    3     3    0   
q-     0     0     1     1     5   0     0     0     0   158    0     0    2   
r-   141   216   242  1003    44   1   162   748   610   350  104    21    1   
s-    94   125   116   886   273   5    64   635  2000   481    1    68    3   
t-   176    25    12  1925    35   1   656   821   364   339   20    88    3   
u-   636   254   684     7   337   0   825  1007   994     2    8     4    1   
v-     1     1     0   227     7   1 






------------------------------------------------------------
Quantitative table for "Asimov_I_Robot"
All chars: 377896


-     -a    -b    -c    -d    -e    -f    -g    -h    -i   -j   -k  \
 -      0  7331  3087  2813  2572  1557  2362  1483  4451  5429  323  435   
a-   1934    19   513  1131   874     9   123   326    38  1153   11  327   
b-     73   363   138     3     1  1337     0     0     0   299   20    0   
c-    280  1331     0   102     3  1215     0     1  1104   446    0  428   
d-   7613   333     5     0   156  1381     9    52    10   926    6    2   
e-  13402  1648    31   795  3448  1043   317   222    62   317   11   50   
f-   2341   387     0     1     0   449   281     0     2   728    0    0   
g-   2188   328     0     0     1   991     0    74   824   346    0    1   
h-   1569  3013     3     0     8  8053     9     0     1  2361    0    0   
i-    925   397   157  1202  1016   819   549   652     0     0    0  195   
j-      0    27     0     0     0    41     0     0     0     8    0    0   
k-    797    41     1     1     4   833     1     5     0   224    0    0   
l-   2197  1092     7    28   757  2022   249     4     1  1215    0   89   
m-   1242  1385   154     6     0  1631    23     0     0   694    0    0   
n-   5461   456    17   658  2961  1850    96  2611    22   735   17  227   
o-   3065    87   981   181   302    95  1913   262    69   352    7  328   
p-    509   517     0     0     3  1064     1     0   180   205    0    0   
q-      1     0     0     0     0     0     0     0     0     0    0    0   
r-   3721  1120   105   200   430  4272    69   162    47  1124    1  171   
s-   7475   775     9   181     3  1938    16     3   935  1113    0  155   
t-   8877  1036     2   139     0  2372    20     1  8643  2110    1    0   
u-   1238   156   168   231   170   208    59   442    21   339    0    0   
v-     12   430     0     0     0  1928     0     0     0   537    0    0   
w-    888  1563     2     0    11  1699     4     0  1118   952    0    4   
x-     57    51     0    51     0    32     0     0     6    85    0    0   
y-   4105    21    26    62    11   522     1     2     1    97    0    1   
z-      8    16     0     0     0   127     0     0     0    24    0    0   

      -l    -m    -n    -o    -p   -q    -r    -s     -t    -u   -v    -w  \
 -  1608  2778  1697  4058  2082  285  2207  5139  10210   850  397  5028   
a-  1990   483  5091     1   394    3  2028  2226   3532   230  619   216   
b-   595     5     7  1192     0    0   382    82     22   580   26     0   
c-   213     0     0  1192     0    4   298    77    691   347    0     0   
d-   129    34   142  1132     2    1   305   195      5   115   36    10   
e-  1771   742  2870    82   383   57  4905  2304   1095    41  588   236   
f-   145     0     0   949     1    0   350     5    213   232    0     3   
g-   276    14    97   475     0    0   379   122     19   121    0     0   
h-    37    39    21  1013     1    0   262    46    536   363    0     1   
i-   999   926  5775  1222   195    8   740  2194   3177    52  453     0   
j-     0     0     0    71     0    0     0     0      0   250    0     0   
k-    47     0   330     4     2    0     0    89      0     5    0    16   
l-  2021    66    28   962    53    0    12   227    187   184  233    26   
m-    27   116    26   628   447    0    76   176      2   239    0     0   
n-   269    32   345  1903    10    3    17   937   2484   132   82    19   
o-   615   910  3666   706   388    2  2451   681   1735  3784  553  1477   
p-   555    11     1  1025   373    0   595   167    141   196    0     9   
q-     0     0     0     0     0    0     0     0      9   382    0     0   
r-   286   255   275  2251   124    0   240   851    761   306   88    31   
s-   239   150   177   869   494   29     0   629   2565   727    1   137   
t-   374    27    21  2502     1    0   718  1119    430   428    1   229   
u-   879   440   842    14   422    0  1257  1160   1477     2   30     0   
v-     0     0     0   178     0    0     7     1      1     4    2     0   
w-






------------------------------------------------------------
Quantitative table for "Harry_Potter_I"
All chars: 412897


-     -a    -b    -c    -d    -e    -f    -g    -h    -i   -j    -k  \
 -      0  7633  3758  2563  2588  1241  2960  1990  8591  4179  348   665   
a-   1905    24   459   920  1450     0   179   970    64  1514    7   490   
b-     48   561   118     0     0  1593     1     0     0   255    7     0   
c-     96   881     0    21     1   760     0   102  1268   141    0  1000   
d-  10483   395     1     3   268  1189     6    73     2   817    0     5   
e-  14249  2195    35   493  4178  1481   315   100   248   364    1    81   
f-   1957   453     2     1     0   685   556     0     2   717    0     0   
g-   3208   431     9     0     0   886     0   145  1097   291    0     0   
h-   1753  4884     3     5    19  9722     6     0     8  3040    0     0   
i-    670   150   140   815  2045   529   427   733     0     0    0   225   
j-      0    31     0     0     0    32     0     0     0     6    0     0   
k-   1368    24     1     0     0  1306    15     1     0   513    0     0   
l-   2080   979    16    66  1011  2926   304     5     0  1285    0   151   
m-   1649   870   389   110     0  1642    21     0     0   853    0     0   
n-   5314   574     9   415  3537  2156    39  3527     2   625    7   302   
o-   3339   271   151   237   407    67  1702   181    78   428    0   695   
p-    822   456    23     0    11  1321     1     0    59   332    0     6   
q-      0     0     0     0     0     0     0     0     0     0    0     0   
r-   4193   925    81   193   612  3998    43   184    30  1656    0   207   
s-   8154  1262    11   248     8  1819    11     9   950   706    0   169   
t-   9159   879     3   317     1  2105    33     0  8794  1084    0     0   
u-    898   118    87   217   415   117   133   597     0   356    0     2   
v-      0    82     0     2     0  2304     0     0     0   330    0     0   
w-    991  2371     6     0    27  1282     2     0  1333  1045    0     8   
x-     37    46     0    59     0    20     0     0     1    21    0     0   
y-   5516    10    26    11    39   612   103     2     3   232    0     3   
z-      2    85     0     0     0   122     0     0     0    25    0     0   

      -l    -m    -n    -o    -p   -q    -r    -s     -t    -u   -v    -w  \
 -  2187  2485  1820  4008  2144  375  1719  6425  11472   938  492  5617   
a-  1845   668  4209     2   563    0  3789  2827   2834   320  557   227   
b-   727     0     1   729     0    0   412    19      7   645   17     0   
c-   484     2     0  1088     0    0   351     3    244   195    0     0   
d-   298    19   448  1250     0    0   336   316      1   483   15    22   
e-  1305   769  2373   178   391    5  5230  1941   1238     3  867   331   
f-   412     0     0  1083     0    0   478     8    241   225    0     8   
g-   226     3    44   989     0    0   933   169      9    85    0    80   
h-    21    16     6  1628     0    0   318    25    695   277    8     8   
i-  1062   998  6213   573   124    2   977  2019   2505    16  378     0   
j-     0     0     0    75     0    0     0     0      0   226    0     0   
k-   113     2   417    10     3    0     0   173      1     2    0     8   
l-  2375    59     4  1336    46    0    44   242    153   210   96    44   
m-    10    97     4   694   254    0   128   221      5   295    0     2   
n-   220     3   158  1559     8    4     5   513   2304   102   46    22   
o-   779  1211  3687  1698   529    0  2668   517   1344  4001  327  1525   
p-   461     6     0   602   401    0   449   123    172   262    0     3   
q-     0     0     0     0     0    0     0     0      0   420    0     0   
r-   183   487   467  2361    69    0  1827  1023    615   245   54    48   
s-   462   153   431  1037   389   34    14   673   2618   368    0   146   
t-   313    47    15  3030     0    0   770   930    714   421   10   259   
u-   929   360  1214     0   547    0  1080  1070   1561     1   33     0   
v-     0     0     0   174     0    0     0     1  






------------------------------------------------------------
Quantitative table for "News_articles"
All chars: 98257


-     -a   -b   -c   -d    -e   -f   -g    -h    -i  -j   -k   -l   -m  \
 -     0  1906  775  897  616   453  880  279   661  1076  81   66  441  707   
a-   564     1  165  285  229     2   73  182     5   288  11  118  641  311   
b-    12   139   11    8    0   299    0    0     0   117   3    0  210    1   
c-    81   405    4   53    0   416    0    0   309   140   0   97   97    3   
d-  1451   141    2    1   44   601    2   32     5   386   3    0   20   23   
e-  3135   430   27  325  818   205  106   89     5   135   4   21  337  200   
f-   547   189    0    0    0   190  110    0     0   196   0    0   99    0   
g-   653   154    0    0    5   285    0   30   110    99   0    0   31    3   
h-   388   672    2    0    1  1630    0    0     2   389   0    0    2    2   
i-    67   265   73  502  318   239  102  193     6     1  36   44  334  195   
j-     1    40    0    0    0    22    0    0     0    11   0    0    0    0   
k-   158    35    0    0    1   189    2    1    33    60   0    2    8    2   
l-   537   345    1    3  153   524   28    0     6   531   0    9  389   14   
m-   247   373   67    6    0   613    1    1     0   309   0    0    0   53   
n-  1496   230    5  266  695   524   50  771     7   247   4   52   32   31   
o-   767    50   39   95   65    15  533   52     8    48   5   38  212  397   
p-   147   235    0    0    1   374    2    0    27    64   0    0  171    4   
q-     1     0    0    0    0     2    0    0     0     4   0    0    0    0   
r-   985   468   42   75   92  1207   11   68     5   515   0   79   69  127   
s-  2488   239    2   62   18   497    5    0   187   385   0   28   38   18   
t-  1448   324    2   14    1   734   25    2  1976   783   0    0   69   18   
u-    75    47   87   58   50   128    9  111     1    42   1    9  181   87   
v-    10    50    0    1    1   554    0    0     0   155   0    0    0    0   
w-   165   181    0    0    7   200    2    0   238   242   0    1    7    7   
x-    42    13    0    7    0    16    0    0     1    17   0    0    0    0   
y-  1011    21    3    6    3   109    1    1     1    25   3    4    4   11   
z-     5     8    0    0    0    48    0    0     6     0   0    0    0    0   

      -n   -o   -p  -q    -r    -s    -t   -u   -v   -w   -x   -y  -z  
 -   386  998  772  32   538  1047  2468  230  114  886    6  155  11  
a-  1268    6  119   1   799   509   851   85  171   42    7  193  35  
b-     0  141    3   0    65    19     6  159    3    0    0  111   0  
c-    14  528    0   1   177    31   190   97    0    0    0   21   1  
d-    23  110    3   0    79    63     0   89   11    1    0   28   0  
e-   925   78  132  13  1309  1012   254   20  165   75  123  128   5  
f-     0  334    0   0   176     2    55   42    0    0    0    2   0  
g-    40  133    1   0   140    61     6   54    0    0    0    7   0  
h-     8  308    0   0    51    16    59   47    0    3    0   19   0  
i-  1603  449   52  11   196   684   688    2  152    1   28    4  20  
j-     1   34    0   0     0     0     0   42    0    0    0    0   0  
k-    23    1    5   0     1    39     1    0    0    1    0    6   0  
l-     5  281   15   0     8   104    73   64   14    5    0  283   0  
m-     1  277  139   0     4    45     0   36    1    0    0   41   0  
n-    56  274    4   0     2   355   682   47   24    5    3   73   1  
o-  1156  130  172   0   869   245   254  515  128  215    6   21   3  
p-     0  267  105   0   259    43    46  111    0    1    1    4   0  
q-     0    0    0   0     0     0     0   55    0    0    0    0   0  
r-   113  498   23   0    86   314   221  143   50    6    1  163   0  
s-    12  182  139   4     2   298   729  162    0   23    0   33   0  
t-     7  699    3   0   273   266    61  105    2   45    0   97   0  
u-   247    3  111   0   290   271   279    0    4    0    8   18   0  
v-     0   65    0   0     0     2     0    2    0    0    0    3   0  
w-    46  155    0   0    27    27   






------------------------------------------------------------
Quantitative table for "Hamlet_BB"
All chars: 211012


-     -a    -b    -c    -d    -e    -f    -g    -h    -i   -j   -k  \
 -     0  4006  1857  1409  1375   828  1598  1008  2786  2918  120  416   
a-   940     0   169   415   454    71   101   246     9   440   25  340   
b-    34   146    22     2     3   813     0     0     1    84   12    0   
c-    60   450     0    38    11   678     0     0   602   137    0  233   
d-  3911   190     0     2    46   864     9    45     2   442    1    1   
e-  7344  1347    26   342   814   764   173    80    31   194    4   45   
f-  1331   355     0     0     3   323   147     0     0   280    0    0   
g-   897   197     0     0    20   486     1    43   400   189    0    0   
h-  1220  1969     2     4    19  3924     5     0     0  1565    0    0   
i-   702   167    49   412   283   460   287   294     1     0    0  141   
j-     2    22     0     0     0    57     0     0     0     3    0    0   
k-   530     3     0     0    22   470     0     0     0   302    0    0   
l-  1381   650     3    29   521  1211   136     5     0   675    0   34   
m-   749   859   103     0    13  1193     3     0     0   267    0    0   
n-  2862   238    24   449  1954   842    62  1208    13   281   10  173   
o-  2179    58    92    88   369   131  1078    53     7   121   16  103   
p-   228   292     1     0    10   492     3     0    92   165    0    1   
q-     0     0     0     0     0     0     0     0     0     0    0    0   
r-  2702   666    29   120   586  1966    29    90    30   649    3   84   
s-  4947   311    15   153    39  1183     8     3   592   505    1   55   
t-  4843   445     4    75     0  1164    21     0  5083   881    0    0   
u-   782    59    67   194   124   214    28   186     0   148    0    3   
v-     0    67     0     0    23  1206     0     0     0   225    0    0   
w-   540   370     0     0    26   607     2     0   860   861    0    1   
x-    23    24     0    34     6    53     1     0     2    46    0    0   
y-  2612    25     0     2     5   173     1     1     0    36    0    0   
z-    38     8     0     0     5    32     0     0     0     3    0    0   

      -l    -m    -n    -o    -p   -q    -r    -s    -t    -u   -v    -w   -x  \
 -  1354  2452  1183  2297  1230  135   684  2831  5849   414  254  2724    0   
a-   889   487  2543     1   194    0  1328  1002  1929   163  459   129   16   
b-   271     0     1   228     0    0   189    39    16   425    1     0    0   
c-   141     1     0   756     0    4   170    11   319   126    0     0    0   
d-    44    16    43   491     0    0   137   284     3    76   19     3    0   
e-   854   378  1837    45   187   22  2560  1471   832    58  256   102  212   
f-    94     0     0   682     1    0   246     3   120   102    0     0    0   
g-    90    16    35   377     0    0   257    79    15   136    0     0    0   
h-    11     9     4  1045     0    0    97    17   293   125    0     1    0   
i-   745   503  2662   528    74    9   487  1699  1533    78  264     1   15   
j-     0     0     0    46     0    0     0     0     0    62    0     0    0   
k-    17     0   157     7     1    0     0    99     0    13    0     1    0   
l-  1301    26     7   863    21    0     6   142   101   102   43     3    0   
m-   178    75    21   497   170    0     3    97     1   179    2     2    0   
n-    39    44   114  1149     7    8     6   477   957    30   40    14    1   
o-   423   695  1608   510   186    2  1848   424   802  2433  221   781    7   
p-   282     1     0   382   123    0   366    64    67   121    0     5    0   
q-     0     0     0     0     0    0     0     0     0   184    0     0    0   
r-    68   137   173   780    55    0   197   609   536   197   55    24    0   
s-    82    39    13   650   257    4    12   393  1371   266    0   118    0   
t-   169    11    10  1384     1    0   414   419   222   226    0   110    0   
u-   512   113   507    19   173    0   984   721   680     0    1     0    2   
v-     0     0






------------------------------------------------------------
Quantitative table for "Sci_articles"
All chars: 325497


-     -a    -b    -c    -d    -e    -f    -g    -h    -i   -j   -k  \
 -      0  6046  1786  3465  2342  1515  2722  1072   960  4343  323  680   
a-   2017    46   606  1193   416    35    50   423    32   678    8  185   
b-    323   418     4    20     0   738     0    18     1   186   61    0   
c-    815  1172     1   186    20  1140     3    67  1393   488    0  227   
d-   4509   504    72    96    50  1988    88    12     0  1007   11    3   
e-  11251  1506    33  1256  2320   794   376   349    31   238   19    9   
f-   2105   146     0    12     3   569   205    10     0   986    0    0   
g-   2331   260    17    54     2   822     3    44   292   365    1    3   
h-   1402  1335     0     2     4  5081     0     3     0   968    0    0   
i-    606   521   172  1363   567   712   481  1044    10    46    0   88   
j-    147    16     0     2     0   105     0     0     0     8    0    1   
k-    605    41     5    29    75   251     0    53    20   193    8    3   
l-   2326  1138    22    97   185  2259    26    90     9  1017    0    7   
m-   1469  1445   166    31     2  1386     5    18     0   533    0    2   
n-   5058   730     1  1059  2388  1208   130  2048    37   726   20   75   
o-   1683   194   301   354   693    59  1918   207    10   197   13   51   
p-    680   625     0    79    19   978     5    11   153   294    1    0   
q-    211     0     0     0    17     7     7     2     0     2    0    0   
r-   3509  1739    67   126   301  4178   232   172     2  1402    0  162   
s-   7758   390    16   378     7  2358   108    16   468  1833   10   63   
t-   4280  1662    25   271    21  3274     8     1  6335  3379    0    3   
u-    186   333   112   293   210   475    31    74     0   157    1   11   
v-    441   588     4    12    37  1717     0     3     6   392    0    0   
w-    468   295     2    11     2   799     0     0   479   821    0    0   
x-    349   119     3    19     1    72     4     2     1   169    2    0   
y-   2273    24    38    14     3   203     0     0     0    81    1    0   
z-    100    37     0     5     0   219     0     0    14    37    0    0   

      -l    -m    -n    -o    -p   -q    -r    -s    -t   -u    -v    -w   -x  \
 -  1297  2106  1098  3349  3115  168  1866  5231  8606  894  1261  2097  107   
a-  2376   986  3609    15   437    5  2025  1528  3690  211   291    24  133   
b-   540     2     4   323    25    0   101    49    62  279     7     0    0   
c-   319    21     4  2091    12    4   285   133  1506  370    55     0    0   
d-    13     4     6   249     4   84   166   220    42  347    26    24    4   
e-  1020   855  2727    48   373  280  4229  2966  1127   46   417   114  478   
f-   132     1     1  1362     2    0   376    11    83  332     0     0    0   
g-    81   132   322   248    95    0   259    42    20  268     0     2    0   
h-    12    96    95   719     3    3   142    14   212   69     0     9    3   
i-   582   920  5496  2119   194   67   357  2112  2006   15   740     1  122   
j-     0     0     0    53    10    0    10     1     0   43    82     0    0   
k-    28     9    40    31    68    0    18    61     3   20     1     5    0   
l-   979    39     2  1013    37    1    43   272   322  369    25    11    2   
m-    23   131    18   751   863    1     6   264    15  290    12     2    0   
n-   167    34   116   712    81    4     5  1223  2755  228   243    15    0   
o-  1206   882  4269   425   590    0  2914   580   799  975   290   828   39   
p-  1131    53     8   860   433    1  1151   100   204  278    16   140    4   
q-     0     0     0     0     2    0     0     6     0  475     0     0    0   
r-    58   542   328  2157    51   97   364   716   554  227    51    28    2   
s-    46    89    18   488   550    8     5   732  2333  526    18    75   10   
t-   219    20    35  1807    87    0  1696   774   332  644     0   163    5   
u-   723   347   652    29   150    5  1224  1071   715    0    15  






------------------------------------------------------------
Quantitative table for "All texts"
All chars: 2263891


-      -a     -b     -c     -d     -e     -f     -g     -h     -i  \
 -      0  43411  17690  19257  14681   9638  16856  10019  24691  28471   
a-  11562    276   3297   6124   4707    280    952   2996    718   6242   
b-   1036   2469    358     40     13   7706      3     19      9   1760   
c-   2372   7287     21    715     41   7201     53    183   8266   2405   
d-  42369   2368     98    120    759   9265    129    295     23   6460   
e-  79473  11005    366   5636  17124   6191   1867   1576    602   2720   
f-  14427   2479      2     21      9   3658   2054     24      4   4152   
g-  13712   2424     34     64     36   6488     12    505   3845   1958   
h-  10502  18077     28     20     94  45248     26      4     24  12640   
i-   4618   2483   1038   7357   6508   4401   3101   4233     63    252   
j-    170    410      1      3      0    384      0      0      0     46   
k-   4642    302     19     31    113   4470     25     66     59   1855   
l-  13499   6647     72    257   4138  14720    960    126     39   7701   
m-   8423   8649   1345    198     22  11791     73     23      3   4390   
n-  32010   4257     81   4293  18986  11040    658  15023    284   4093   
o-  17262    944   2298   1732   3133    576  12614   1393    425   1581   
p-   3343   3588     42    106     52   6856     16     14   1478   1652   
q-    229      6      0      0     17     12      7      2      1      8   
r-  23823   8216    383    978   2993  25057    480   1151    156   8521   
s-  51841   4668     91   1577     88  13383    195    113   4787   7469   
t-  41282   6563     94   1037     30  15714    157     18  50334  14066   
u-   4601   1284    888   1987   1397   1746    332   2158     29   1409   
v-    530   1820      4     15     61  12326      2      3      6   2810   
w-   4673   6504     15     45     98   6493     22      1   6203   6380   
x-   1187    428     32    287     10    241      5     14     24    436   
y-  23506    148    236    139     80   2470    108    186      7    731   
z-    219    258      1      5      6    860      1      0     22    208   

      -j    -k     -l     -m     -n     -o     -p    -q     -r     -s     -t  \
 -  2121  3059  10513  16610   9263  24615  15370  1162  10588  30310  61470   
a-    98  2080  13634   5119  27631    135   2735    23  15403  12045  20611   
b-   139     1   3439     14     47   3821     30     0   2280    442    170   
c-     0  2674   1932     28     21   9134     14    20   2215    483   4608   
d-    23    11    667    151   1009   4604     23    85   1668   1848     61   
e-    47   310   8886   4734  18902   1046   2516   635  28192  17631   7256   
f-     0     1   1234      1      1   7167     24     0   2661    111   1078   
g-     1     4    924    222    803   3591     97     5   2991    789    153   
h-     0     5    163    208    238   7340     16     3   2102    244   2589   
i-    37  1085   6260   5267  34002   8238   1019   145   4347  15540  15180   
j-     1     1      0      0      1    626     13     1     12      2      0   
k-     8    10    318     14   1288     77     95     0     53    689     14   
l-     0   400  11627    296     61   6975    254     1    150   2020   1399   
m-     1     4    265    698    125   5360   2742     1    245   1296     27   
n-    96  1275   1064    257   1116   8393    127    26     63   6001  14299   
o-    69  1716   5278   8905  22284   4531   3153     5  16974   5155   7866   
p-     1    10   4461    135     10   4555   2109     1   5136    731   1122   
q-     0     0      0      0      1      1      7     0      0      6     10   
r-     5   978    944   2663   2115  13075    514    98   3330   5633   4143   
s-    39   642   1021    885    794   6107   2829   121    196   4599  13966   
t-     2     3   1514    183    116  14715    131     1   5702   5474   2500   
u-     5    88   4968   2072   5276     96   2108     6   7331   6748   6712   
v-     0     0     10     25     23   

---
---
# Creating ***frequency*** and ***biased*** tables

We got dict with DataFrames wich contain values of number of pairs for each text files and all texts.

For further work, we interesting only the table for `'All texts'`.

---
## Creating ***frequency*** tables
Now we will create dict with dataframes wich contain:
- values of number pairs for all texts
- values of number pairs for all texts without spaces($^1$)
- frequency values of number pairs for all texts ($^2$)
- frequency values of number pairs for all texts without spaces ($^1$)($^2$)

---
($^1$) Окрім стандартної кількісної табліці (літери та пробіл), додатково створюємо таблицю тільки для літер. Це пов'язане з тим, що кодований текст немає пробілів.

($^2$) Варто зауважити, що для ціх датафреймів (матриць) виникають проблеми з точністю.  Це проявляється в тому сенсі, що сума елементів частотних значень (нормочаних матриць) не дорівнює 1, а наприклад 1.0000000000000002. Тобто відрізняється від 1 на значення приблизно 1е-15.
В даному випадку в нас не так багато букв (2263891) в тестовому прикладі, тому це неповинно вплинути на результат. Але якщо робити частотну таблицю по набагато більш об'ємному набору текстів, то можуть виникнкти проблеми. (хоча і не факт). [Дивись прикдад в кінці ноутбуку `Example 1`]

In [14]:
freq_tables = {}

# DF for values of number pairs for all texts
freq_tables['Quantit for All'] = quantit_tables['All texts'].copy()
freq_tables['Quantit for All'].name = 'Quantit for All'

# DF for values of number pairs for all texts without spaces
freq_tables['Quantit without spaces'] = freq_tables['Quantit for All'].iloc[1:, 1:]
freq_tables['Quantit without spaces'].name = 'Quantit without spaces'

# DF for frequency of number pairs for all texts
freq_tables['Freq for All'] = freq_tables['Quantit for All'] / freq_tables['Quantit for All'].values.sum()
freq_tables['Freq for All'].name = 'Freq for All'

# DF for frequency of number pairs for all texts without spaces
freq_tables['Freq without spaces'] = freq_tables['Quantit without spaces'] / freq_tables['Quantit without spaces'].values.sum()
freq_tables['Freq without spaces'].name = 'Freq without spaces'


Save DataFrames with `quantitative` and `frequency` tables for `All texts` to disk (Goodle drive)

In [15]:
for name in freq_tables:
    path_pt = os.path.join(drch, folder_pt, name + '.csv')
    freq_tables[name].to_csv(path_pt, index_label=False)

---
## Creating ***biased*** tables

Now we create dict with ***Biased tables***.
Biased tables is a tables with bias, that is, we change the values in the table that are equal to 0 to 1.

Як можна бачити, пари букв з малим значенням кільності (1-3) доволі "дивні". Це викликано якоюсь аномалією в тексті: або слово дивного походження, або помилка. Тому ми можемо припустити, що всі "аномалії" зустрілися хоча б 1 раз, тобто зробити заміну кількості з 0 на 1.

Now we create follow dataframes that contain `biased tables` by `frequency tables`, namely:
- biased quantitativity table for all texts
- biased quantitativity table for all texts without spaces
- biased frequency table for all texts
- biased frequency table for all texts without spaces

In [16]:
biased_tables = {}

# DF for biased values (BVal) of number pairs for all texts
biased_tables['BQuant for All'] = (quantit_tables['All texts'].copy()).replace(0, 1)
biased_tables['BQuant for All'].name = 'BQuant for All'

# DF for values of number pairs for all texts without spaces
biased_tables['BQuant without spaces'] = biased_tables['BQuant for All'].iloc[1:, 1:]
biased_tables['BQuant without spaces'].name = 'BQuant without spaces'

# DF for frequency of number pairs for all texts
biased_tables['BFreq for All'] = biased_tables['BQuant for All'] / biased_tables['BQuant for All'].values.sum()
biased_tables['BFreq for All'].name = 'BFreq for All'

# DF for frequency of number pairs for all texts without spaces
biased_tables['BFreq without spaces'] = biased_tables['BQuant without spaces'] / biased_tables['BQuant without spaces'].values.sum()
biased_tables['BFreq without spaces'].name = 'BFreq without spaces'

Save this DataFrames with `biased tables` to dick.

In [17]:
for name in biased_tables:
    path_pt = os.path.join(drch, folder_pt, name + '.csv')
    biased_tables[name].to_csv(path_pt, index_label=False)

## Creating ***logarithmic*** tables
We got biased tables, but they contain values, which when multiplied many times give result, that more (quantitative) or less (frequency) then conputer know. Therefore, to avoid large (or small) values, we need make a `logarithmic` operation (on base `10`).

Now we create follow dataframes that contain `logarithmic tables` by `biased tables`, namely:
- logarithmic biased quantitativity table for all texts
- logarithmic biased quantitativity table for all texts without spaces
- logarithmic biased frequency table for all texts
- logarithmic biased frequency table for all texts without spaces

In [18]:
logarithmic_tables = {}

# DF for logarithmic biased values (Log BVal) of number pairs for all texts
logarithmic_tables['Log BQuant for All'] = biased_tables['BQuant for All'].applymap(np.log10)
logarithmic_tables['Log BQuant for All'].name = 'Log BQuant for All'

# DF for logarithmic biased values of number pairs for all texts without spaces
logarithmic_tables['Log BQuant without spaces'] = biased_tables['BQuant without spaces'].applymap(np.log10)
logarithmic_tables['Log BQuant without spaces'].name = 'Log BQuant without spaces'

# DF for logarithmic biased frequency of number pairs for all texts
logarithmic_tables['Log BFreq for All'] = biased_tables['BFreq for All'].applymap(np.log10)
logarithmic_tables['Log BFreq for All'].name = 'Log BFreq for All'

# DF for logarithmic biased frequency of number pairs for all texts without spaces
logarithmic_tables['Log BFreq without spaces'] = biased_tables['BFreq without spaces'].applymap(np.log10)
logarithmic_tables['Log BFreq without spaces'].name = 'Log BFreq without spaces'

In [19]:
for name in logarithmic_tables:
    path_pt = os.path.join(drch, folder_pt, name + '.csv')
    logarithmic_tables[name].to_csv(path_pt, index_label=False)

---
---
---